### DISCUSSION TITLE
How to generate a list of distinct (non-overlapping) time periods

### DISCUSSION THREAD

**Discussion by Spookyguy** [instancio/instancio#856] [Q&A]
I am trying to generate a list of objects with distinct time periods. I have a POJO like this:

```java
public class TimePeriod {
    private ZonedDateTime begin;
    private ZonedDateTime end;
}
```

I want to generate a list of those objects like so:

```java
Instancio.ofList(TimePeriod.class)
// ???
.create();
``` 

where the time periods do not overlap. I had the idea to use a custom generator, but according to the documentation, it is not possible to get hold of already generated objects to generate a time period that does not reach into already generated ones.

Any ideas?

**Comment by armandino** [comment]
There's an example similar to what you described in #843. Does that work?

**Comment by Spookyguy** [comment]
Yes, that's  the workaround i'm currently using:

```java
final AtomicReference<ZonedDateTime> dateHolder = new AtomicReference<>(ZonedDateTime.now().plusDays(1).truncatedTo(ChronoUnit.DAYS));
final List<TimePeriod> timePeriods = IntStream.range(0, 30).mapToObj(i -> Instancio.of(TimePeriod.class)
  .generate(field(TimePeriod::getBegin), g -> g.temporal().zonedDateTime().range(dateHolder.get().plusHours(8), dateHolder.getAndUpdate(d -> d.plusDays(1)).plusHours(20)))
  .assign(Assign.valueOf(field(TimePeriod::getBegin)).to(field(TimePeriod::getEnd)).as((ZonedDateTime dt) -> dt.plusHours(2)))
  .create()).toList();
``` 

But it creates only one TimePeriod per day. A generator for generating distinct TimePeriods would be way more flexible and i would not have to use the IntStream. BTW: My real object is a little bit more complex (Something like an appointment containing a begin and an end and some other attributes).

**Reply by armandino** [reply]
Thanks for the additional info. It's true that a generator isn't aware of values generated outside itself. In this case, we could make it stateful to keep track of the last generated date:

```java
class TimePeriodGenerator implements Generator<TimePeriod> {

    private ZonedDateTime nextStart = ZonedDateTime.now().plusDays(1);

    @Override
    public TimePeriod generate(final Random random) {
        final long periodLengthMinutes = random.longRange(20, 60);
        final long durationBetweenPeriodMinutes = random.longRange(1, 10);

        TimePeriod period = new TimePeriod();
        period.setBegin(nextStart);
        period.setEnd(nextStart.plusMinutes(periodLengthMinutes));

        nextStart = period.getEnd().plusMinutes(durationBetweenPeriodMinutes);
        return period;
    }
}
```

This should create non-overlapping time periods, and the rest of the fields should be random values: 

```java
List<TimePeriod> periods = Instancio.ofList(TimePeriod.class)
    .size(10)
    .supply(all(TimePeriod.class), new TimePeriodGenerator())
    .create();
```

If you register the generator via the [GeneratorProvider SPI](https://www.instancio.org/user-guide/#instancio-service-provider-interface), then it will be picked up automatically, without doing:

```java
supply(all(TimePeriod.class), new TimePeriodGenerator())
```

**Reply by Spookyguy** [reply]
Very cool. Didn't think of that a generator could be stateful. Thanks for the hint. Should have thought of and tried it myself :D ...
